eclipse-collections-api/src/main/java/org/eclipse/collections/api/RichIterable.java

```diff
@@ -147,6 +147,30 @@ default T getOnly()
 
         throw new IllegalStateException("Size must be 1 but was " + this.size());
     }
+    
+    /**
+     * Returns the element as an optional if the iterable has exactly one element.
+     * If the iterable is empty {@link Optional#empty} is returned.
+     * Otherwise, throw {@link IllegalStateException}.
+     *
+     * @return an element of an iterable wrapped in an optional or an empty optional.
+     * @throws IllegalStateException if iterable has more than one element.
```

Comment:
Very interesting points here. I think we would all agree that RichIterable.getOnlyOptional() would return empty() for an empty collection and a present Optional for a single element. I had assumed empty() for two or more elements but I suppose that's not the only obvious choice. Iterables.getOnlyElement() looks exactly the same as Iterate.getOnly(), in that it throws for both empty and two or more. MoreCollectors.toOptional() has Javadoc which says `The returned collector throws an IllegalArgumentException if the stream consists of two or more elements, and a NullPointerException if the stream consists of exactly one element, which is null.` That last part means they delegate to Optional.of() instead of Optional.ofNullable() which is crucial. It's the only way to differentiate between empty and a collection containing one null.

It seems like this API could use more discussion, it's more complex than it seemed when we created the issue.